From 37597e0d28548c02fb4ae0f558a70781ae9feba2 Mon Sep 17 00:00:00 2001
From: Xinyu Chen <xinyu.chen@freescale.com>
Date: Mon, 15 Nov 2010 16:19:16 +0800
Subject: [PATCH 429/670] ENGR00123559 Android lowmemorykiller judges the
 number of free pages incorrect.

low memory killer has problem to judge the current free pages in system.
It gets total reclaimable anon cache and free pages as it's judging data.
But on our platforms, we have DMA zone there, and normal process can not
allocate memory from DMA. So here, just remove the DMA's
free/anon pages from this data. Let low memory killer works well.

Signed-off-by: Xinyu Chen <xinyu.chen@freescale.com>
---
 drivers/staging/android/lowmemorykiller.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/drivers/staging/android/lowmemorykiller.c b/drivers/staging/android/lowmemorykiller.c
index d663faa..4235bc6 100644
--- a/drivers/staging/android/lowmemorykiller.c
+++ b/drivers/staging/android/lowmemorykiller.c
@@ -17,6 +17,7 @@
  * and processes may not get killed until the normal oom killer is triggered.
  *
  * Copyright (C) 2007-2008 Google, Inc.
+ * Copyright (C) 2010 Freescale Semiconductor, Inc.
  *
  * This software is licensed under the terms of the GNU General Public
  * License version 2, as published by the Free Software Foundation, and
@@ -35,6 +36,8 @@
 #include <linux/oom.h>
 #include <linux/sched.h>
 #include <linux/notifier.h>
+#include <linux/nodemask.h>
+#include <linux/vmstat.h>
 
 static uint32_t lowmem_debug_level = 2;
 static int lowmem_adj[6] = {
@@ -73,6 +76,15 @@ static int lowmem_shrink(struct shrinker *s, int nr_to_scan, gfp_t gfp_mask)
 	int array_size = ARRAY_SIZE(lowmem_adj);
 	int other_free = global_page_state(NR_FREE_PAGES);
 	int other_file = global_page_state(NR_FILE_PAGES);
+	int node;
+
+	for_each_node_state(node, N_HIGH_MEMORY) {
+		struct zone *z =
+			&NODE_DATA(node)->node_zones[ZONE_DMA];
+
+		other_free -= zone_page_state(z, NR_FREE_PAGES);
+		other_file -= zone_page_state(z, NR_FILE_PAGES);
+	}
 
 	/*
 	 * If we already have a death outstanding, then
-- 
1.8.0

